### ISSUE TITLE
cross entropy gives nan

### ISSUE THREAD

**Issue by bellonemauro** [tiny-dnn/tiny-dnn#410]
Dear Developer,

first of all thanks for sharing with the community your amazing work. 
I have recently started to use this library in one of my project and I have noted some numerical instability using the cross-entropy loss function with relu activation, sometimes I get NaN in some neurons so the network drops. 
I am aware of the possible numerical instability exists in the theory, however, is there any function to prevent the instability or at least warn the user that the train did not succeed ? 

Does anybody experimented this before?

Thanks
Mauro

**Comment by beru** [comment]
@bellonemauro If you could kindly provide minimum example (source and data) to reproduce the problem,  That would be a big help.

**Comment by pi-null-mezon** [comment]
I have also faced with descripted "Nan problem" before. And in my expirience the best practice to prevent weights blow is to use double precision for weights (defined in config.h). Also you can use tan_h activation instead of relu, learning to the same loss values can take more epoch but Nan problem will not bother you any more.

But, if you strongly want to use relu activation, to control Nan weights condition you can run training with following lambda, which will be executed after each learning epoch:
````c++
m_net.train<cross_entropy>(_opt, tvects, tlbls, static_cast<size_t>(_minibatch), _epoch,
                                    [](){ /* Executed after each minibatch*/},
                                    [&](){
                                            static int epoch = 0;
                                            if(((epoch % 1) == 0) || (epoch == _epoch - 1)) {
                                                if(clbls.size() > 0) {
                                                    tiny_cnn::result cresult = m_net.test(cvects,clbls);
                                                    tiny_cnn::result tresult = m_net.test(tvects,tlbls);                                                   
                                                    std::cout << "  " << epoch << "\t" << tresult.accuracy() << " / " << cresult.accuracy() << std::endl;
                                                } else {
                                                    tiny_cnn::result tresult = m_net.test(tvects,tlbls);
                                                    std::cout << " " << epoch << "\t" << tresult.accuracy() << " / unknown" << std::endl;
                                                }
                                            }
                                            epoch++;
                                          });
````
When Nan weights appear in the network, accuracy on training and control sets drops near to 50 %.

**Comment by bellonemauro** [comment]
Hi,
thanks for your answer, I have tried using double, it is more robust but it doesn't solve, I will prepare a minimal example so the error will be further documented. 
I can confirm that my net works using other configurations and activation functions. 

Moreover, I must say that even with the same data I can run the training several times and I do not always get nan, and I also get different performances (between 80% and 95%). I believe that the variations in the performances depends on the random weights initialization, it should converge increasing epochs and number of neurons. 
When it goes to 50% it just means that it is not working as it picks classes randomly.

However, what I think is that there should be some nan check in the activation function to detect when a neuron is failing in order to give an exception because that neuron cannot be activated anymore and the error will propagate. 
This can be an improvement in the robustness of the library.

**Comment by edgarriba** [comment]
@bellonemauro notice that some parts of the activation functions code have no robust unit tests
https://coveralls.io/builds/8872207/source?filename=tiny_dnn%2Factivations%2Factivation_function.h

A contribution here could be very appreciated.

**Comment by Gammalyse** [comment]
`d += -t[i] * std::log(y[i]) - (float_t(1) - t[i]) * std::log(float_t(1) - y[i]);`

My money is on -inf through log(0) and nan through -inf multiplied by zero. That would happen if the output y[i] would be perfectly 1.

**Comment by MatthewScholefield** [comment]
Disclaimer: I have very little knowledge on the inner workings of cross entropy loss and adagrad/adam

I found that I was getting `nan` whenever my inputs had `0` in it. However, after changing 0 to something like 0.01, running `fit()` with cross entropy just makes the output values get closer and closer to either -1 or +1 despite some that should be going to 0.01.

**Comment by elpidiovaldez** [comment]
I had problems with cross_entropy loss going to NaN and setting weights to NaN.  I tracked this to y[i] reaching 1.0 precisely when t[i] was 1.0. This causes the loss and gradients to give arithmetic errors.  Now when output and target are 1 or 0, the loss should be 0.  I fixed the problem by changing the loss to:

```
// cross-entropy loss function for (multiple independent) binary classifications
class cross_entropy {
 public:
  static float_t f(const vec_t &y, const vec_t &t) {
    assert(y.size() == t.size());
    float_t d{0};

    for (size_t i = 0; i < y.size(); ++i)
      d += -t[i] * std::log(y[i]) -
           (float_t(1) - t[i]) * std::log(float_t(1) - y[i]);

    if(d!=d) d=0; //Catch NaN when t[i]==y[i]==1 or t[i]==y[i]==0

    return d;
  }

  static vec_t df(const vec_t &y, const vec_t &t) {
    assert(y.size() == t.size());
    vec_t d(t.size());

    for (size_t i = 0; i < y.size(); ++i) {
      d[i] = (y[i] - t[i]) / (y[i] * (float_t(1) - y[i]));
      if(d[i]!=d[i]) d[i]=0; //Catch NaN when t[i]==y[i]==1 or t[i]==y[i]==0
    }

    return d;
  }
};


```

**Comment by samathasolupati** [comment]
> I had problems with cross_entropy loss going to NaN and setting weights to NaN. The problem was quite hard to trace because when NaNs start to appear they rapidly propagate causing weights, losses and gradients all to become NaN. The problem arises in BOTH the cross-entropy loss AND gradient functions.
> 
> My issue was due to an output probability, y[i] reaching 1.0 precisely. The target value was also 1.0 so in the loss, the expression (1-t[i])_log(1-y[i]) evaluates to 0_log(0) which returns NaN. Plainly the loss should be 0 when target and predicted values are both 1.0. A similar issue arises when target value and predicted value are both exactly 0.0. Again the desired loss is 0.
> 
> The gradient of the loss suffers from a divide-by-0 error when the predicted value, y[i]==0.0 or 1.0. This is seen by examining the expression (y[i] - t[i]) / (y[i] * (1 - y[i])). It is not completely clear to me what the limit of this expression is as y[i] -> t[i] for t[i]==0 and t[i]==1. Numerically it appears to be -1 for t[i]==1 and 1 for t[i]==0.
> 
> I modified the cross_entropy loss as shown below and my problems with Nan disappeared.
> 
> ```
> // cross-entropy loss function for (multiple independent) binary classifications
> class cross_entropy {
>  public:
>   static float_t f(const vec_t &y, const vec_t &t) {
>     assert(y.size() == t.size());
>     float_t d{0};
> 
>     for (size_t i = 0; i < y.size(); ++i)
>       d += -t[i] * std::log(y[i]) -
>            (float_t(1) - t[i]) * std::log(float_t(1) - y[i]);
> 
>     if(d!=d) d=0; //Catch NaN when t[i]==y[i]==1 or t[i]==y[i]==0
> 
>     return d;
>   }
> 
>   static vec_t df(const vec_t &y, const vec_t &t) {
>     assert(y.size() == t.size());
>     vec_t d(t.size());
> 
>     for (size_t i = 0; i < y.size(); ++i) {
>       d[i] = (y[i] - t[i]) / (y[i] * (float_t(1) - y[i]));
>       if(d[i]!=d[i]) d[i] = t[i]==0 ? 1.0 : -1.0; //Catch NaN when t[i]==y[i]==1 or t[i]==y[i]==0
>     }
> 
>     return d;
>   }
> };
> ```
What we should do for multi class cross entropy? Iam getting nan with multiclass cross entropy
